### PR TITLE
resolve: known-answer suppression fix for ageing entries

### PR DESCRIPTION
**Pull request by vchillara**
Optimization for known answer suppression for ageing entries.
RFC 6762 section 7.1 suggests that a Multicast DNS querier SHOULD NOT include records in the Known-Answer list whose remaining TTL is less than half of their original TTL.
This is useful in general to keep the information in cache current, but particularly so when browsing for mDNS services (PR #22532 ).
Skipping cached records whose age is >50% TTL from the known answers enables us to receive updated info from the network if the service is still alive.

### CODE REVIEW

File: src/resolve/resolved-dns-cache.c [hunk 1]
```diff
@@ -1265,6 +1266,14 @@ int dns_cache_export_shared_to_packet(DnsCache *cache, DnsPacket *p) {
                         if (!j->shared_owner)
                                 continue;
 
+                        if (t <= 0)
+                                t = now(CLOCK_BOOTTIME);
```
Comment (medhefgo) [inline src/resolve/resolved-dns-cache.c]: This should be done outside the loop.

File: src/resolve/resolved-dns-cache.c [hunk 2]
```diff
@@ -1265,6 +1266,14 @@ int dns_cache_export_shared_to_packet(DnsCache *cache, DnsPacket *p) {
...
+
+                        /* RFC6762 7.1: Don't append records with less than
+                         * half the TTL remaining as known answers. */
+                        if ((t >= j->until) || (j->until - t) < (0.5 * (j->rr->ttl * USEC_PER_SEC)))
```
Comment (medhefgo) [inline src/resolve/resolved-dns-cache.c]: I doubt thing this needs float math:
```suggestion
                        if ((t >= j->until) || (j->until - t) < (j->rr->ttl * (USEC_PER_SEC / 2)))
```

File: src/resolve/resolved-dns-cache.c (hunk 1 above)
Comment (vchillara) [reply src/resolve/resolved-dns-cache.c]: Moved outside

File: src/resolve/resolved-dns-cache.c (hunk 2 above)
Comment (vchillara) [reply src/resolve/resolved-dns-cache.c]: Changed as suggested

**Comment by vchillara** [comment]
@medhefgo addressed your comments. Any further feedback?

File: src/resolve/resolved-dns-cache.c [hunk 3]
```diff
@@ -1265,6 +1268,11 @@ int dns_cache_export_shared_to_packet(DnsCache *cache, DnsPacket *p) {
                         if (!j->shared_owner)
                                 continue;
 
+                        /* RFC6762 7.1: Don't append records with less than
+                         * half the TTL remaining as known answers. */
+                        if ((t >= j->until) || (j->until - t) < (j->rr->ttl * (USEC_PER_SEC / 2)))
```
Comment (yuwata) [inline src/resolve/resolved-dns-cache.c]: Please drop redundant `()`, and use `usec_sub_unsigned()`:
`if (usec_sub_unsigned(t, j->until) < j->rr->ttl * USEC_PER_SEC / 2)`

File: src/resolve/resolved-dns-cache.c [hunk 4]
```diff
@@ -1265,6 +1268,11 @@ int dns_cache_export_shared_to_packet(DnsCache *cache, DnsPacket *p) {
                         if (!j->shared_owner)
                                 continue;
 
+                        /* RFC6762 7.1: Don't append records with less than
+                         * half the TTL remaining as known answers. */
```
Comment (yuwata) [inline src/resolve/resolved-dns-cache.c]: Not a big matter, but it is not necessary to wrap comment so frequently. We allow about 109ch per line.

File: src/resolve/resolved-dns-cache.c [hunk 5]
```diff
@@ -1252,11 +1252,14 @@ int dns_cache_check_conflicts(DnsCache *cache, DnsResourceRecord *rr, int owner_
 int dns_cache_export_shared_to_packet(DnsCache *cache, DnsPacket *p) {
```
Comment (yuwata) [inline src/resolve/resolved-dns-cache.c:1252]: But please wrap the long line in the commit message.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/resolve/resolved-dns-cache.c [hunk 6]
```diff
@@ -1265,6 +1268,12 @@ int dns_cache_export_shared_to_packet(DnsCache *cache, DnsPacket *p) {
                         if (!j->shared_owner)
                                 continue;
 
+                        /* RFC6762 7.1: Don't append records with less than half the TTL remaining
+                         * as known answers.
+                         */
```
Comment (yuwata) [inline src/resolve/resolved-dns-cache.c]: Not a fun of unnecessary line break at the end of comment. But, not a big matter.

**Review by yuwata — APPROVED** [review]
LGTM.

**Pull request merged**